src/linesensor.cpp

```diff
@@ -0,0 +1,163 @@
+#include <cmath>
+#include <esp_log.h>
+
+#include "linesensor.h"
+
+#define TAG "Mcp3008LineSensor"
+
+namespace mcp3008 {
+
+LineSensor::LineSensor() : m_installed(false), m_spi(NULL), m_spi_dev(HSPI_HOST) {
+
+}
+
+LineSensor::~LineSensor() {
+    uninstall();
+}
+
+esp_err_t LineSensor::install(const LineSensor::Config& cfg) {
+    if(m_installed)
+        return ESP_OK;
+
+    esp_err_t ret;
+    spi_bus_config_t buscfg = { 0 };
+    buscfg.miso_io_num = cfg.pin_miso;
+    buscfg.mosi_io_num = cfg.pin_mosi;
+    buscfg.sclk_io_num = cfg.pin_sck;
+    buscfg.quadwp_io_num = -1;
+    buscfg.quadhd_io_num = -1;
+
+    spi_device_interface_config_t devcfg = { 0 };
+    devcfg.clock_speed_hz = cfg.freq;
+    devcfg.mode = 0;
+    devcfg.spics_io_num = cfg.pin_cs;
+    devcfg.queue_size = CHANNELS;
+
+    ret = spi_bus_initialize(cfg.spi_dev, &buscfg, 1);
+    if(ret != ESP_OK) {
+        return ret;
+    }
+
+    ret = spi_bus_add_device(cfg.spi_dev, &devcfg, &m_spi);
+    if(ret != ESP_OK) {
+        spi_bus_free(cfg.spi_dev);
+        return ret;
+    }
+
+    m_spi_dev = cfg.spi_dev;
+    m_installed = true;
+    return ESP_OK;
+}
+
+esp_err_t LineSensor::uninstall() {
+    if(!m_installed)
+        return ESP_OK;
+
+    esp_err_t res = spi_bus_remove_device(m_spi);
+    if(res != ESP_OK)
+        return res;
+
+    res = spi_bus_free(m_spi_dev);
+    if(res != ESP_OK)
+        return res;
+
+    m_installed = false;
+    return ESP_OK;
+}
+
+esp_err_t LineSensor::read(std::vector<uint16_t>& results, bool differential, uint8_t channels_mask) {
+    int requested = 0;
+    for(int i = 0; i < CHANNELS; ++i) {
+        if(((1 << i) & channels_mask) != 0)
+            ++requested;
+    }
+
+    const size_t orig_size = results.size();
+    results.resize(orig_size + requested);
+
+    esp_err_t res = this->read(results.data() + orig_size, differential, channels_mask);
+    if(res != ESP_OK) {
+        results.resize(orig_size);
+        return res;
+    }
+    return ESP_OK;
+}
+
+esp_err_t LineSensor::read(uint16_t *dest, bool differential, uint8_t channels_mask) {
+    if(!m_installed)
+        return ESP_FAIL;
+
+    int requested = 0;
+    spi_transaction_t transactions[CHANNELS] = { 0 };
+    for(int i = 0; i < CHANNELS; ++i) {
+        if(((1 << i) & channels_mask) == 0)
+            continue;
+
+        auto &t = transactions[i];
+        t.user = (void*)requested;
+        t.flags = SPI_TRANS_USE_RXDATA | SPI_TRANS_USE_TXDATA;
+        t.length = 3 * 8;
+        t.tx_data[0] = 1;
+        t.tx_data[1] = (differential << 7) | ((i & 0x07) << 4);
```

Comment:
Note to self: bug, `(0 << 7)` means differential and `(1 << 7)` means single-ended, it's the other way around here.